examples/vision/simsiam.py

```diff
@@ -0,0 +1,434 @@
+between two different versions of the same image. The hope is that in this way the model
+will have learn representations where the similar images are grouped as together possible
+while the dissimilar images are further away. 
+
+In this example, we will be implementing one such system called **SimSiam** proposed in
+[Exploring Simple Siamese Representation Learning](https://arxiv.org/abs/2011.10566). It
+is implemented as the following:
+1. We create two different versions of the same dataset with a stochastic data
+augmentation pipeline. Note that the random initialization seed needs to be the same
+during create these versions. 
+2. We take a ResNet without any classification head (**backbone**) and we add a shallow
+fully-connected network (**projection head**) on top of it. Collectively, this is known
+as the **encoder**. 
+3. We pass the output of the encoder through a **predictor** which is again a shallow
+fully-connected network having an
+[AutoEncoder](https://en.wikipedia.org/wiki/Autoencoder) like structure. 
+3. We then train our encoder to maximize the cosine similarity between the two different
+versions of our dataset. 
+
+This example requires TensorFlow 2.4 or higher.
+"""
+
+"""
+## Setup
+"""
+
+from tensorflow.keras import layers
+from tensorflow.keras import regularizers
+import tensorflow as tf
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+"""
+## Define hyperparameters
+"""
+
+AUTO = tf.data.AUTOTUNE
+BATCH_SIZE = 128
+EPOCHS = 5
+CROP_TO = 32
+SEED = 26
+
+PROJECT_DIM = 2048
+LATENT_DIM = 512
+WD = 0.0005
+
+"""
+## Load the CIFAR10 dataset
+"""
+
+(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()
+print(f"Total training examples: {len(x_train)}")
+print(f"Total test examples: {len(x_test)}")
+
+"""
+## Defining our data augmentation pipeline
+
+As studied in [SimCLR](https://arxiv.org/abs/2002.05709) having the right data
+augmentation pipeline is critical for SSL systems to work effectively in computer vision.
+Two particular augmentation transforms that seem to matter the most are: 1.) Random
+resized crops and 2.) Color distortions. Most of the other SSL systems for computer
+vision (such as [BYOL](https://arxiv.org/abs/2006.07733),
+[MoCoV2](https://arxiv.org/abs/2003.04297), [SwAV](https://arxiv.org/abs/2006.09882),
+etc.) include these in their training pipelines.  
+"""
+
+
+def flip_random_crop(image):
+    # With random crops we also apply horizontal flipping.
+    image = tf.image.random_flip_left_right(image)
+    image = tf.image.random_crop(image, (CROP_TO, CROP_TO, 3))
+    return image
+
+
+def color_jitter(x, s=[0.4, 0.4, 0.4, 0.1]):
+    x = tf.image.random_brightness(x, max_delta=0.8 * s[0])
+    x = tf.image.random_contrast(x, lower=1 - 0.8 * s[1], upper=1 + 0.8 * s[1])
+    x = tf.image.random_saturation(x, lower=1 - 0.8 * s[2], upper=1 + 0.8 * s[2])
+    x = tf.image.random_hue(x, max_delta=0.2 * s[3])
```

Comment:
Notable thoughts. Thank you! 

But as studied in SimCLR and all the works that followed it affine transforms and crops are collectively responsible to prevent the representation collapse situation.